Zend/zend_enum.c

```diff
@@ -371,3 +376,128 @@ void zend_enum_register_props(zend_class_entry *ce)
 		zend_declare_typed_property(ce, ZSTR_KNOWN(ZEND_STR_VALUE), &value_default_value, ZEND_ACC_PUBLIC, NULL, value_type);
 	}
 }
+
+static const zend_function_entry simple_internal_methods[] = {
+	ZEND_NAMED_ME(cases, zend_enum_cases_func, arginfo_class_UnitEnum_cases, ZEND_ACC_PUBLIC | ZEND_ACC_STATIC)
+	ZEND_FE_END
+};
+
+static const zend_function_entry backed_internal_methods[] = {
+	ZEND_NAMED_ME(from, zend_enum_from_func, arginfo_class_BackedEnum_from, ZEND_ACC_PUBLIC | ZEND_ACC_STATIC)
+	ZEND_NAMED_ME(tryFrom, zend_enum_try_from_func, arginfo_class_BackedEnum_tryFrom, ZEND_ACC_PUBLIC | ZEND_ACC_STATIC)
+	ZEND_FE_END
+};
+
+ZEND_API zend_class_entry *zend_register_internal_enum(const char *name)
+{
+	return zend_register_internal_backed_enum_ex(name, IS_UNDEF, NULL);
+}
+
+ZEND_API zend_class_entry *zend_register_internal_enum_ex(const char *name, const zend_function_entry *functions)
+{
+	return zend_register_internal_backed_enum_ex(name, IS_UNDEF, functions);
+}
+
+ZEND_API zend_class_entry *zend_register_internal_backed_enum(const char *name, uint32_t backing_type)
+{
+	return zend_register_internal_backed_enum_ex(name, backing_type, NULL);
+}
+
+ZEND_API zend_class_entry *zend_register_internal_backed_enum_ex(const char *name, uint32_t backing_type, const zend_function_entry *functions)
+{
+	ZEND_ASSERT(backing_type == IS_UNDEF || backing_type == IS_STRING || backing_type == IS_LONG);
+
+	zend_class_entry ce_val, *ce;
+	INIT_CLASS_ENTRY_EX(ce_val, name, strlen(name), functions);
+	ce = zend_register_internal_class(&ce_val);
+
+	ce->ce_flags |= ZEND_ACC_ENUM;
+	ce->enum_backing_type = backing_type;
+	ce->backed_enum_table = malloc(sizeof(HashTable));
+	zend_hash_init(ce->backed_enum_table, 0, NULL, ZVAL_PTR_DTOR, 1);
+	zend_enum_register_props(ce);
+
+	zend_register_functions(ce, simple_internal_methods, &ce->function_table, EG(current_module)->type);
+	zend_do_implement_interface(ce, zend_ce_unit_enum);
+	if (backing_type != IS_UNDEF) {
+		zend_register_functions(ce, backed_internal_methods, &ce->function_table, EG(current_module)->type);
+		zend_do_implement_interface(ce, zend_ce_backed_enum);
+	}
+
+	return ce;
+}
+
+static zend_object *add_internal_enum_case(zend_class_entry *ce, zend_string *name, zval *value)
+{
+	zend_object *case_obj = zend_objects_new_persistent(ce);
+	enum_obj_init(case_obj, name, value);
+
+	zval casezv;
+	ZVAL_OBJ(&casezv, case_obj);
+	zend_class_constant *c = zend_declare_class_constant_ex(ce, name, &casezv, ZEND_ACC_PUBLIC, NULL);
+	ZEND_CLASS_CONST_FLAGS(c) |= ZEND_CLASS_CONST_IS_CASE;
+
+	zend_string_release(name);
```

Comment:
I feel it's unergonomic to do `zend_string * ... = zend_string_init(..); zend_register_enum_case(...); zend_string_release(...);` it essentially becomes three statements instead of just one `zend_register_enum_case(... zend_string_init (..))` … I particular I cannot imagine a case where you would want to retain the zend_string afterwards.